src/Microsoft.ML.TimeSeries/SrCnnBatchAnomalyDetection.cs

```diff
@@ -0,0 +1,778 @@
+
+        private readonly int _batchSize;
+        private readonly string _inputColumnName;
+        private readonly int _outputLength;
+        private readonly SrCnnEntireModeler _modler;
+
+        private class Bindings : ColumnBindingsBase
+        {
+            private readonly VectorDataViewType _outputColumnType;
+            private readonly int _inputColumnIndex;
+
+            public Bindings(DataViewSchema input, string inputColumnName, string outputColumnName, VectorDataViewType outputColumnType)
+                : base(input, true, outputColumnName)
+            {
+                _outputColumnType = outputColumnType;
+                _inputColumnIndex = Input[inputColumnName].Index;
+            }
+
+            protected override DataViewType GetColumnTypeCore(int iinfo)
+            {
+                Contracts.Check(iinfo == 0);
+                return _outputColumnType;
+            }
+
+            // Get a predicate for the input columns.
+            public Func<int, bool> GetDependencies(Func<int, bool> predicate)
+            {
+                Contracts.AssertValue(predicate);
+
+                var active = new bool[Input.Count];
+                for (int col = 0; col < ColumnCount; col++)
+                {
+                    if (!predicate(col))
+                        continue;
+
+                    bool isSrc;
+                    int index = MapColumnIndex(out isSrc, col);
+                    if (isSrc)
+                        active[index] = true;
+                    else
+                        active[_inputColumnIndex] = true;
+                }
+
+                return col => 0 <= col && col < active.Length && active[col];
+            }
+        }
+
+        public SrCnnBatchAnomalyDetector(IHostEnvironment env, IDataView input, string inputColumnName, string outputColumnName, double threshold, int batchSize, double sensitivity, SrCnnDetectMode detectMode)
+            : base(env, "SrCnnBatchAnomalyDetector", input)
+        {
+            Contracts.CheckValue(env, nameof(env));
+
+            Contracts.CheckValue(inputColumnName, nameof(inputColumnName));
+            _inputColumnName = inputColumnName;
+
+            env.CheckUserArg(batchSize == -1 || batchSize >= MinBatchSize, nameof(batchSize), "BatchSize must be -1 or no less than 12.");
+            if (batchSize == -1)
+            {
+                _batchSize = (int)input.GetRowCount();
+            }
+            else
+            {
+                _batchSize = batchSize;
+            }
+
+            env.CheckUserArg(threshold >= 0 && threshold <= 1, nameof(threshold), "Must be in [0,1].");
+            env.CheckUserArg(detectMode == SrCnnDetectMode.AnomalyOnly
+                || detectMode == SrCnnDetectMode.AnomalyAndExpectedValue
+                || detectMode == SrCnnDetectMode.AnomalyAndMargin, nameof(detectMode), "Invalid detectMode");
+
+            if (detectMode.Equals(SrCnnDetectMode.AnomalyOnly))
```

Comment:
>AnomalyOnly [](start = 50, length = 11)

Since this is an enum, this can be done using a `switch` statement.

Reply:
As commented above, you don't need an if or a switch statement if you store the possible values for _outputLength in an const array whose indices match the the enum values.

---
In reply to: [426408554](https://github.com/dotnet/machinelearning/pull/5135#discussion_r426408554) [](ancestors = 426408554)

Reply:
Fixed.

---
In reply to: [426828074](https://github.com/dotnet/machinelearning/pull/5135#discussion_r426828074) [](ancestors = 426828074,426408554)